File: packages/python/examples/mcp_modern_roundtrip.py
"""Exercise the modern MCP client through a real in-process HTTP loopback."""

import json
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer

from axllm import AxMCPClient, AxMCPStreamableHTTPTransport


class Handler(BaseHTTPRequestHandler):
    calls = 0
    tool_lists = 0
    failures = []

    def log_message(self, *args):
        pass

    def do_POST(self):
        length = int(self.headers.get("Content-Length", "0"))
        request = json.loads(self.rfile.read(length) or b"{}")
        method = request.get("method")
        params = request.get("params") or {}
        if method == "initialize":
            self.failures.append("modern client sent initialize")
        if method != "server/discover" and "_meta" not in params:
            self.failures.append(f"{method} omitted request _meta")
        Handler.calls += 1
        meta = {"io.modelcontextprotocol/serverInfo": {"name": "modern-loopback", "version": f"1.0.{self.calls}"}}
        result = {"resultType": "complete", "_meta": meta}
        if method == "server/discover":
            result = {"resultType": "complete", "supportedVersions": ["2026-07-28"], "capabilities": {"tools": {}, "extensions": {"io.modelcontextprotocol/tasks": {}}}, "ttlMs": 60000, "cacheScope": "public", "_meta": meta}
        elif method == "tools/list":
            Handler.tool_lists += 1
            result.update({
                "tools": [
                    {"name": "start_reindex", "inputSchema": {"type": "object", "properties": {"scope": {"type": "string", "x-mcp-header": "Scope"}}}},
                    {"name": "mrtr_roots_round", "inputSchema": {"type": "object", "properties": {}}},
                ],
                "ttlMs": 60000,
                "cacheScope": "public",
            })
        elif method == "tools/call" and params.get("name") == "start_reindex":
            if self.headers.get("Mcp-Param-Scope") != "all":
                self.failures.append("Mcp-Param-Scope was not propagated")
            result = {"resultType": "task", "taskId": "task-1", "status": "working", "createdAt": "2026-07-29T00:00:00Z", "lastUpdatedAt": "2026-07-29T00:00:00Z", "ttlMs": None, "_meta": meta}
        elif method == "tasks/get":
            result = {"taskId": "task-1", "status": "completed", "createdAt": "2026-07-29T00:00:00Z", "lastUpdatedAt": "2026-07-29T00:00:01Z", "ttlMs": None, "result": {"resultType": "complete", "structuredContent": {"indexed": 42}, "_meta": meta}, "_meta": meta}
        elif method == "tools/call" and "requestState" not in params:
            result = {"resultType": "input_required", "inputRequests": {"roots": {"method": "roots/list"}}, "requestState": "opaque-roots-state", "_meta": meta}
        elif method == "tools/call":
            if params.get("requestState") != "opaque-roots-state" or params.get("inputResponses", {}).get("roots", {}).get("roots", [{}])[0].get("uri") != "file:///workspace":
                self.failures.append("roots MRTR response was not echoed")
            result.update({"structuredContent": {"roots": 1}})
        payload = json.dumps({"jsonrpc": "2.0", "id": request.get("id"), "result": result}).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


server = HTTPServer(("127.0.0.1", 0), Handler)
threading.Thread(target=server.serve_forever, daemon=True).start()
try:
    transport = AxMCPStreamableHTTPTransport(
        f"http://127.0.0.1:{server.server_address[1]}/mcp",
        {"ssrfProtection": {"requireHttps": False, "allowLocalhost": True, "allowPrivateNetworks": True}},
    )
    client = AxMCPClient(transport, {"era": "modern", "roots": [{"uri": "file:///workspace", "name": "workspace"}]})
    client.init()
    assert client.get_era() == "modern"
    client.refresh(force=False)
    task = client.call_tool("start_reindex", {"scope": "all"})
    assert task.get("structuredContent", {}).get("indexed") == 42, task
    roots = client.call_tool("mrtr_roots_round", {})
    assert roots.get("structuredContent", {}).get("roots") == 1, roots
    catalog = client.inspect_catalog()
    assert Handler.tool_lists == 1, "cacheable tools/list was fetched again"
    assert not Handler.failures, Handler.failures
    assert catalog.get("serverInfo", {}).get("version") != "1.0.1", catalog.get("serverInfo")
    client.close()
finally:
    server.shutdown()

print("mcp-modern-roundtrip-ok")

File: packages/rust/examples/mcp_modern_roundtrip.rs
use axllm::{AxMCPClient, AxMCPStreamableHTTPTransport, AxResult};
use serde_json::{json, Value};
use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream};
use std::thread;
use std::time::Duration;

fn read_request(stream: &mut TcpStream) -> String {
    stream.set_read_timeout(Some(Duration::from_secs(5))).ok();
    let mut request = Vec::new();
    let mut tmp = [0u8; 4096];
    let mut expected = None;
    loop {
        let read = stream.read(&mut tmp).unwrap_or(0);
        if read == 0 {
            break;
        }
        request.extend_from_slice(&tmp[..read]);
        if expected.is_none() {
            if let Some(end) = request.windows(4).position(|value| value == b"\r\n\r\n") {
                let headers = String::from_utf8_lossy(&request[..end + 4]);
                let length = headers
                    .lines()
                    .find_map(|line| {
                        line.to_ascii_lowercase()
                            .strip_prefix("content-length:")
                            .map(|value| value.trim().parse::<usize>().unwrap_or(0))
                    })
                    .unwrap_or(0);
                expected = Some(end + 4 + length);
            }
        }
        if expected.is_some_and(|length| request.len() >= length) {
            break;
        }
    }
    String::from_utf8_lossy(&request).into_owned()
}

fn main() -> AxResult<()> {
    let listener = TcpListener::bind("127.0.0.1:0").expect("bind loopback");
    let port = listener.local_addr().unwrap().port();
    let server = thread::spawn(move || {
        let mut calls = 0usize;
        let mut tool_lists = 0usize;
        let mut failures = Vec::new();
        loop {
            let (mut stream, _) = listener.accept().expect("accept loopback");
            let raw = read_request(&mut stream);
            let body = raw.split("\r\n\r\n").nth(1).unwrap_or("{}");
            let request: Value = serde_json::from_str(body).expect("JSON-RPC request");
            let method = request["method"].as_str().unwrap_or_default();
            let params = request.get("params").cloned().unwrap_or_else(|| json!({}));
            if method == "initialize" {
                failures.push("modern client sent initialize".to_string());
            }
            if method != "server/discover" && params.get("_meta").is_none() {
                failures.push(format!("{method} omitted request _meta"));
            }
            calls += 1;
            let meta = json!({"io.modelcontextprotocol/serverInfo":{"name":"modern-loopback","version":format!("1.0.{calls}")}});
            let mut result = json!({"resultType":"complete","_meta":meta});
            let mut done = false;
            match method {
                "server/discover" => {
                    result = json!({"resultType":"complete","supportedVersions":["2026-07-28"],"capabilities":{"tools":{},"extensions":{"io.modelcontextprotocol/tasks":{}}},"ttlMs":60000,"cacheScope":"public","_meta":meta})
                }
                "tools/list" => {
                    tool_lists += 1;
                    result = json!({"resultType":"complete","tools":[
                        {"name":"start_reindex","inputSchema":{"type":"object","properties":{"scope":{"type":"string","x-mcp-header":"Scope"}}}},
                        {"name":"mrtr_roots_round","inputSchema":{"type":"object","properties":{}}}
                    ],"ttlMs":60000,"cacheScope":"public","_meta":meta});
                }
                "tools/call" if params["name"] == "start_reindex" => {
                    if !raw.to_ascii_lowercase().contains("mcp-param-scope: all") {
                        failures.push("Mcp-Param-Scope was not propagated".into());
                    }
                    result = json!({"resultType":"task","taskId":"task-1","status":"working","createdAt":"2026-07-29T00:00:00Z","lastUpdatedAt":"2026-07-29T00:00:00Z","ttlMs":null,"_meta":meta});
                }
                "tasks/get" => {
                    result = json!({"taskId":"task-1","status":"completed","createdAt":"2026-07-29T00:00:00Z","lastUpdatedAt":"2026-07-29T00:00:01Z","ttlMs":null,"result":{"resultType":"complete","structuredContent":{"indexed":42},"_meta":meta},"_meta":meta})
                }
                "tools/call" if params.get("requestState").is_none() => {
                    result = json!({"resultType":"input_required","inputRequests":{"roots":{"method":"roots/list"}},"requestState":"opaque-roots-state","_meta":meta})
                }
                "tools/call" => {
                    if params["requestState"] != "opaque-roots-state"
                        || params["inputResponses"]["roots"]["roots"][0]["uri"]
                            != "file:///workspace"
                    {
                        failures.push("roots MRTR response was not echoed".into());
                    }
                    result = json!({"resultType":"complete","structuredContent":{"roots":1},"_meta":meta});
                    done = true;
                }
                _ => failures.push(format!("unexpected method {method}")),
            }
            let envelope = json!({"jsonrpc":"2.0","id":request["id"],"result":result}).to_string();
            let response = format!("HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}", envelope.len(), envelope);
            stream.write_all(response.as_bytes()).unwrap();
            stream.flush().unwrap();
            if done {
                break;
            }
        }
        (tool_lists, calls, failures)
    });

    let transport = AxMCPStreamableHTTPTransport::new(
        format!("http://127.0.0.1:{port}/mcp"),
        json!({"ssrfProtection":{"requireHttps":false,"allowLocalhost":true,"allowPrivateNetworks":true}}),
    )?;
    let mut client = AxMCPClient::new(
        Box::new(transport),
        json!({"era":"modern","roots":[{"uri":"file:///workspace","name":"workspace"}]}),
    );
    client.init()?;
    assert_eq!(client.get_era(), Some("modern"));
    client.refresh_with_force(false)?;
    let task = client.call_tool("start_reindex", json!({"scope":"all"}))?;
    assert_eq!(task["structuredContent"]["indexed"], 42);
    let roots = client.call_tool("mrtr_roots_round", json!({}))?;
    assert_eq!(roots["structuredContent"]["roots"], 1);
    let catalog = client.inspect_catalog(false)?;
    client.close()?;
    let (tool_lists, calls, failures) = server.join().unwrap();
    assert_eq!(tool_lists, 1, "cacheable tools/list was fetched again");
    assert!(
        calls >= 6 && failures.is_empty(),
        "modern roundtrip failures: {failures:?}"
    );
    assert_ne!(
        catalog.server_info["version"], "1.0.1",
        "serverInfo did not refresh"
    );
    println!("mcp-modern-roundtrip-ok");
    Ok(())
}
